File: app/services/postprocess.py
"""Transcription post-process filter.

Single source of truth for the decision: is this Whisper output noise or content?
Used by /listen, /transcribe, /ask, and the OpenAI-compat endpoints so the
empty/punctuation/silence rules stay aligned across surfaces.

This module is pure — no logging, no I/O, no config reads. Callers (HTTP/WS
handlers) read config and emit `transcription_filtered` log lines on Drop.

The empty-text check strips:
  - All Unicode whitespace via `str.isspace()`.
  - A hand-rolled set of ASCII + CJK punctuation. We avoid the `regex` library
    (not in deps) and `unicodedata.category(c) == "P*"` (would also drop
    apostrophes inside English words like "don't"). The hand-rolled set covers
    the Whisper hallucination cases we actually see; extend it when new
    failure modes surface.
"""

from dataclasses import dataclass
from typing import Literal

_PUNCT_CHARS: frozenset[str] = frozenset(
    ".,!?;:'\"`-_*~/\\()[]{}<>"  # ASCII punctuation Whisper emits
    "。，、；：？！「」『』（）《》〈〉…—·"  # CJK punctuation Whisper emits
)


@dataclass(frozen=True)
class Keep:
    text: str


@dataclass(frozen=True)
class Drop:
    reason: Literal["empty_text", "below_min_duration"]


FilterDecision = Keep | Drop


def filter_empty_transcription(
    text: str,
    duration_ms: float | None,
    *,
    enabled: bool,
    min_duration_ms: int,
) -> FilterDecision:
    """Decide whether a transcription result is content or noise.

    Args:
        text: Raw text from the WhisperBackend.
        duration_ms: Source audio duration in milliseconds, or None when the
            caller did not measure it (e.g. /transcribe). When None the
            duration check is skipped.
        enabled: When False, always returns `Keep(text)` (kill-switch).
        min_duration_ms: Minimum acceptable utterance duration. Must be >= 0.

    Returns:
        `Keep(text)` if the transcription is content, or `Drop(reason)` if it
        should be suppressed. `Keep.text` is the ORIGINAL text — no stripping.

    Raises:
        ValueError: if `min_duration_ms` is negative.
    """
    if min_duration_ms < 0:
        raise ValueError(f"min_duration_ms must be non-negative; got {min_duration_ms}")

    if not enabled:
        return Keep(text)

    if duration_ms is not None and duration_ms < min_duration_ms:
        return Drop("below_min_duration")

    if _is_empty_after_stripping(text):
        return Drop("empty_text")

    return Keep(text)


def _is_empty_after_stripping(text: str) -> bool:
    """Return True when text contains nothing but whitespace and punctuation."""
    for ch in text:
        if ch.isspace():
            continue
        if ch in _PUNCT_CHARS:
            continue
        return False
    return True

File: engine/core/src/postprocess.rs
//! Transcription post-process filter. Pure port of
//! `app/services/postprocess.py` — single source of truth for
//! "is this Whisper output noise or content?".

/// ASCII + CJK punctuation Whisper emits in hallucinated/noise output.
/// Mirrors `_PUNCT_CHARS` in the Python module exactly.
const PUNCT_CHARS: &str = ".,!?;:'\"`-_*~/\\()[]{}<>。，、；：？！「」『』（）《》〈〉…—·";

#[derive(Debug, PartialEq, Eq)]
pub enum FilterDecision {
    Keep(String),
    Drop(DropReason),
}

#[derive(Debug, PartialEq, Eq)]
pub enum DropReason {
    EmptyText,
    BelowMinDuration,
}

impl DropReason {
    pub fn as_str(&self) -> &'static str {
        match self {
            DropReason::EmptyText => "empty_text",
            DropReason::BelowMinDuration => "below_min_duration",
        }
    }
}

/// Decide whether a transcription result is content or noise.
/// `duration_ms = None` skips the duration check (the /transcribe path).
pub fn filter_empty_transcription(
    text: &str,
    duration_ms: Option<f64>,
    enabled: bool,
    min_duration_ms: u64,
) -> FilterDecision {
    if !enabled {
        return FilterDecision::Keep(text.to_owned());
    }
    if let Some(d) = duration_ms {
        if d < min_duration_ms as f64 {
            return FilterDecision::Drop(DropReason::BelowMinDuration);
        }
    }
    if is_empty_after_stripping(text) {
        return FilterDecision::Drop(DropReason::EmptyText);
    }
    FilterDecision::Keep(text.to_owned())
}

fn is_empty_after_stripping(text: &str) -> bool {
    text.chars()
        .all(|c| c.is_whitespace() || PUNCT_CHARS.contains(c))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn keeps_real_content() {
        assert_eq!(
            filter_empty_transcription("你好 world", None, true, 500),
            FilterDecision::Keep("你好 world".into())
        );
    }

    #[test]
    fn drops_pure_punctuation_and_whitespace() {
        for noise in ["", "   ", "。。。", " .,!? ", "「」…—·", "\n\t"] {
            assert_eq!(
                filter_empty_transcription(noise, None, true, 500),
                FilterDecision::Drop(DropReason::EmptyText),
                "should drop {noise:?}"
            );
        }
    }

    #[test]
    fn apostrophe_inside_word_is_kept() {
        // "don't" must survive — the Python module explicitly avoids
        // stripping apostrophes that carry content.
        assert_eq!(
            filter_empty_transcription("don't", None, true, 500),
            FilterDecision::Keep("don't".into())
        );
    }

    #[test]
    fn duration_check_applies_only_when_measured() {
        assert_eq!(
            filter_empty_transcription("hi", Some(100.0), true, 500),
            FilterDecision::Drop(DropReason::BelowMinDuration)
        );
        assert_eq!(
            filter_empty_transcription("hi", None, true, 500),
            FilterDecision::Keep("hi".into())
        );
    }

    #[test]
    fn kill_switch_keeps_everything() {
        assert_eq!(
            filter_empty_transcription("。", None, false, 500),
            FilterDecision::Keep("。".into())
        );
    }
}
